service/controller/resource/app/delete.go

```diff
@@ -0,0 +1,138 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/k8smetadata/pkg/label"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/v6/pkg/controller/context/finalizerskeptcontext"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/cluster-apps-operator/pkg/project"
+	"github.com/giantswarm/cluster-apps-operator/service/controller/key"
+)
+
+func (r Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var apps []*v1alpha1.App
+	{
+		r.logger.Debugf(ctx, "finding apps for cluster '%s/%s'", cr.GetNamespace(), key.ClusterID(&cr))
+
+		// Get all apps for cluster not being managed by cluster-apps-operator.
+		selector, err := labels.Parse(fmt.Sprintf("%s=%s,%s!=%s", label.Cluster, key.ClusterID(&cr), label.ManagedBy, project.Name()))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		o := client.ListOptions{
+			Namespace:     cr.GetNamespace(),
+			LabelSelector: selector,
+		}
+
+		var appList v1alpha1.AppList
+
+		err = r.ctrlClient.List(ctx, &appList, &o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, item := range appList.Items {
+			apps = append(apps, item.DeepCopy())
+		}
+
+		r.logger.Debugf(ctx, "found %d apps for cluster '%s/%s'", len(apps), cr.GetNamespace(), key.ClusterID(&cr))
+	}
+
+	if len(apps) > 0 {
+		r.logger.Debugf(ctx, "waiting for %d apps to be deleted for cluster '%s/%s'", len(apps), cr.GetNamespace(), key.ClusterID(&cr))
```

Comment:
The deletion logic changes a lot. We first check if there are any apps remaining for the cluster. Once all apps are gone we delete chart-operator and then app-operator.

This is because once app-operator has been deleted any app CRs will stay in terminating due to finalizers. 

However if the customer doesn't delete the apps the cluster CR will stay in terminating. I created https://github.com/giantswarm/giantswarm/issues/20449 to add a metric for this. Thanks for the idea @tfussell